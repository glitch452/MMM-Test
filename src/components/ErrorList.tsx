import React, { JSX } from 'jsx-dom-cjs';

export interface LoadingErrorsProps {
  title?: string;
  error_list: string[];
}

const ErrorList = ({ title, error_list }: LoadingErrorsProps): JSX.Element => {
  return (
    <div className="loading small">
      {title}
      {title && error_list.length > 0 && <br />}
      {error_list.map((e, i) => (
        <>
          {e}
          {i !== error_list.length - 1 && <br />}
        </>
      ))}
    </div>
  );
};

export default ErrorList;
