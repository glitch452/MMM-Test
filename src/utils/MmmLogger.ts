import Log from 'logger';

/**
 * A helper class to assist with logging.
 * It logs to the MagicMirror logger by default, but an alternate transport can be provided for it to log to.
 * It also decides whether or not to log, based on the provided log level.
 */
export default class MmmLogger implements Module.Logger {
  /**
   * The cut-off level for logging.
   * Lower-level logs will be logged, higher-level logs will no be.
   */
  private _level: Module.LoggerLevels = Module.LoggerLevels.ERROR;

  /**
   * The cut-off level for logging.
   * Lower-level logs will be logged, higher-level logs will no be.
   */
  get level() {
    return this._level;
  }

  /**
   * The module properties, used to inject the module name into the logs.
   */
  private readonly _properties: Module.ModulePropertiesExt<unknown>;

  /**
   * The module properties, used to inject the module name into the logs.
   */
  get properties() {
    return this._properties;
  }

  /**
   * The underlying logger where the logs are sent.
   */
  readonly transport: typeof Log;

  /**
   * Create a new instance.
   * @param properties The module properties, used to inject the module name into the logs.
   * @param level The cut-off level for logging.
   * @param transport The underlying logger where the logs are sent.
   */
  constructor(
    properties: Module.ModulePropertiesExt<unknown>,
    level: Module.LoggerLevels | Module.LoggerLevel = Module.LoggerLevels.ERROR,
    transport: typeof Log = Log,
  ) {
    this._properties = properties;
    this.setLogLevel(level);
    this.transport = transport;
  }

  /**
   * Change the current level of the logger.
   * @param level The cut-off level for logging.
   */
  setLogLevel(level: Module.LoggerLevels | Module.LoggerLevel) {
    if (typeof level === 'string') {
      this._level = this.convertLevel(level);
    } else {
      this._level = level;
    }
  }

  /**
   * Used to log non-categorized log messages.
   * @param message The string to be logged
   */
  log(message: string) {
    this.transport.log(this.processMessage(message));
  }

  /**
   * Used to log messages in the info category.
   * @param message The string to be logged
   */
  info(message: string) {
    this.transport.info(this.processMessage(message));
  }

  /**
   * Used to log messages in the warning category.
   * @param message The string to be logged
   */

  warn(message: string) {
    if (this._level >= Module.LoggerLevels.WARN) {
      this.transport.warn(this.processMessage(message));
    }
  }

  /**
   * Used to log messages in the error category.
   * @param message The string to be logged
   */
  error(message: string) {
    if (this._level >= Module.LoggerLevels.ERROR) {
      this.transport.error(this.processMessage(message));
    }
  }

  /**
   * Used to log messages in the debug category.
   * @param message The string to be logged
   */
  debug(message: string) {
    if (this._level >= Module.LoggerLevels.DEBUG) {
      this.transport.info(this.processMessage(message));
    }
  }

  /**
   * Formats the raw log message by adding the module name.
   * For debug messages, the time and module index are also added.
   * @param message The string to be logged
   * @returns The formatted log string
   */
  private processMessage(message: string): string {
    if (this._level === Module.LoggerLevels.DEBUG) {
      const date = new Date();
      const time = `${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}`;
      return `${this._properties.name}: (${this._properties.data.index})(${time}) ${message}`;
    }
    return `${this._properties.name}: ${message}`;
  }

  /**
   * Converts the level name to the enum representation.
   * @throws an Error if an invalid level name is provided
   * @param level_name The name of the log level to be converted
   * @returns The enum representation of the requested level
   */
  private convertLevel(level_name: Module.LoggerLevel): Module.LoggerLevels {
    switch (level_name) {
      case 'INFO':
        return Module.LoggerLevels.INFO;
      case 'WARN':
        return Module.LoggerLevels.WARN;
      case 'ERROR':
        return Module.LoggerLevels.ERROR;
      case 'DEBUG':
        return Module.LoggerLevels.DEBUG;
      default:
        throw new Error('Invalid Logger Level');
    }
  }
}
