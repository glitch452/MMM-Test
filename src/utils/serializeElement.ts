import { JSDOM } from 'jsdom';
import { JSX } from 'jsx-dom-cjs';

const WINDOW = global.window;
const DOCUMENT = global.document;
const NAVIGATOR = global.navigator;

const DOM = new JSDOM('<!doctype html><html><body></body></html>');

/**
 * Sets up the global vars required for creating DOM elements
 */
function setDomVars() {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  global.window = DOM.window as unknown as typeof global.window;
  global.document = DOM.window.document;
  global.navigator = DOM.window.navigator;
}

setDomVars();

/**
 * @param element A DOM element to snapshot
 * @returns The HTML string representation of the element
 */
export function serializeElement(element: JSX.Element | (() => JSX.Element)): string {
  setDomVars();

  return typeof element === 'function' ? element().outerHTML : element.outerHTML;
}

/**
 * @param element A DOM element to snapshot
 * @returns The HTML string representation of the element
 */
export function serializeElementSandbox(element: () => JSX.Element): string {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  global.window = DOM.window as unknown as typeof global.window;
  global.document = DOM.window.document;
  global.navigator = DOM.window.navigator;

  const snapshot: string = element().outerHTML;

  global.window = WINDOW;
  global.document = DOCUMENT;
  global.navigator = NAVIGATOR;

  return snapshot;
}
