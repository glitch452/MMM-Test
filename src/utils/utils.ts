/**
 * @param input The template string
 * @param find The substring to find within the input string
 * @param replace The value to replace the find string with
 * @returns The input string with the find values replaced by the replace value
 */
export function replaceAll(input: string, find: string, replace: string): string {
  if (find === replace) {
    return input;
  }
  let processing = input;
  let output = '';
  let idx = input.indexOf(find);
  while (idx >= 0) {
    output += processing.substring(0, idx) + replace;
    processing = processing.substring(idx + find.length);
    idx = processing.indexOf(find);
  }
  output += processing;
  return output;
}
