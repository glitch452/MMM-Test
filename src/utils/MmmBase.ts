import { ZodSchema } from 'zod';
import MmmLogger from './MmmLogger';

export interface MinimumConfig {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  logLevel: Module.LoggerLevel;
}

type This = Module.ModuleProperties<MinimumConfig>;

export type MmmBase = Pick<
  This,
  'init' | 'setConfig' | 'start' | 'suspend' | 'resume' | 'notificationReceived'
>;

export interface MmmBaseConstructor {
  new <O extends MinimumConfig, D, I>(schema: ZodSchema<O, D, I>, logger?: Module.Logger): MmmBase;
}

// eslint-disable-next-line @typescript-eslint/consistent-type-assertions, @typescript-eslint/naming-convention
export const MmmBase = function <O extends MinimumConfig, D, I>(
  this: This,
  schema: ZodSchema<O, D, I>,
  logger?: Module.Logger,
) {
  /**
   * Initialize standard and module specific fields
   *
   * @param this Magic Mirror Module object
   */
  this.init = function (this: This) {
    this.has_config_error = false;
    this.config_errors = [];
    this.logger = logger ?? new MmmLogger(this);
    this.requiresVersion = '2.1.0';
  };

  /**
   * Override the setConfig function to use zod for parsing the configuration values
   *
   * @param this Magic Mirror Module object
   * @param config The user specified configuration options
   */
  this.setConfig = function (this: This, config: unknown): void {
    const result = schema.safeParse(config);
    this.has_config_error = !result.success;

    if (result.success) {
      this.config = result.data;
      this.logger.setLogLevel?.(this.config.logLevel);
    } else {
      for (const ze of result.error.errors) {
        const message = `'${ze.path}': ${ze.message}`;
        this.config_errors.push(message);
        this.logger.error(`Configuration error '${ze.code}' in option ${message}`);
      }
    }
  };

  this.start = function (this: This) {
    this.logger.debug(`start(): this.data: ${JSON.stringify(this.data)}`);
    this.logger.debug(`start(): this.config: ${JSON.stringify(this.config)}`);
  };

  this.suspend = function () {
    this.logger.info(this.translate('SUSPENDING'));
  };

  this.resume = function () {
    this.logger.info(this.translate('RESUMING'));
  };

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  this.notificationReceived = function () {};
} as unknown as MmmBaseConstructor;
