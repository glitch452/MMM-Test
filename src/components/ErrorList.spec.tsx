import { expect } from 'chai';
import ErrorList from './ErrorList';
import React from 'jsx-dom-cjs';
import { serializeElement } from 'src/utils/serializeElement';
import 'chai-snapshot-matcher';

describe('LoadingErrors', function () {
  it('should render with no errors', function () {
    const actual = serializeElement(<ErrorList error_list={[]} />);
    expect(actual).to.matchSnapshot(this);
  });

  it('should render with no errors and a title', function () {
    const actual = serializeElement(<ErrorList title="test" error_list={[]} />);
    expect(actual).to.matchSnapshot(this);
  });

  it('should render with 1 error', function () {
    const actual = serializeElement(<ErrorList title="test" error_list={['error one']} />);
    expect(actual).to.matchSnapshot(this);
  });

  it('should render with more than 1 error', function () {
    const actual = serializeElement(<ErrorList error_list={['error one', 'error two']} />);
    expect(actual).to.matchSnapshot(this);
  });
});
