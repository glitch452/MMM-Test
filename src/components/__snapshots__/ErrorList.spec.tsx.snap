exports["LoadingErrors should render with no errors"] = "<div class=\"loading small\"></div>";

exports["LoadingErrors should render with no errors and a title"] = "<div class=\"loading small\">test</div>";

exports["LoadingErrors should render with 1 error"] = "<div class=\"loading small\">test<br>error one</div>";

exports["LoadingErrors should render with more than 1 error"] = "<div class=\"loading small\">error one<br>error two</div>";

